src/graph/topo.cc

```diff
@@ -1036,6 +1041,73 @@ ncclResult_t scclGetAllAlgoFromXMLFilesAndSetComm(struct ncclComm* comm, const c
   return ncclSuccess;
 }
 
+ncclResult_t scclGetAllAlgoFromSCCLConfigAndSetComm(struct ncclComm* comm, const char* str){
+  INFO(NCCL_INIT, "SCCL: Parsing config %s", str);
+  struct ncclXml* xml;
+
+  comm->scclRegistrations = NULL;
+  comm->nScclRegistrations = 0;
+
+  NCCLCHECK(ncclCalloc(&xml, 1));
+  NCCLCHECK(scclGetXmlConfigFromFile(str, xml));
+
+  struct ncclXmlNode* topNode;
+  NCCLCHECK(xmlFindTag(xml, "sccl_algos", &topNode));
+
+  for (int s=0; s < topNode->nSubs; s++) {
+    struct ncclXmlNode* node = topNode->subs[s];
+    if (strcmp(node->name, "load") == 0) {
+      if (comm->numberOfSCCAlgorithms == SCCL_MAX_NUM_ALGOS){
+        WARN("SCCL: too many algorithms (%d) specified in environment variable SCCL_XML_FILES. The rest will be ignored.", comm->numberOfSCCAlgorithms);
+        break;
+      }
+
+      const char *path;
+      NCCLCHECK(xmlGetAttrStr(node, "path", &path));
+
+      int hasMinsize = false;
+      NCCLCHECK(xmlAttrExists(node, "minsize", &hasMinsize));
+      int64_t minsize = 0;
+      if (hasMinsize) {
+        NCCLCHECK(xmlGetAttrInt64_t(node, "minsize", &minsize));
+      }
+
+      int hasMaxsize = false;
+      NCCLCHECK(xmlAttrExists(node, "maxsize", &hasMaxsize));
+      int64_t maxsize = -1; // Represents infinity
+      if (hasMaxsize) {
+        NCCLCHECK(xmlGetAttrInt64_t(node, "maxsize", &maxsize));
+      }
+
+      int hasProtocol = false;
+      NCCLCHECK(xmlAttrExists(node, "proto", &hasProtocol));
+      const char *protocol = NULL;
+      if (hasProtocol) {
+        NCCLCHECK(xmlGetAttrStr(node, "proto", &protocol));
+      }
+
+      int algoIndex = comm->numberOfSCCAlgorithms;
+      struct scclAlgorithm* scclAlgo = &comm->scclAlgos[algoIndex];
+      if (scclGetAlgoFromXMLAndSetComm(comm, path, scclAlgo) == ncclSuccess){
+        comm->numberOfSCCAlgorithms++;
+        INFO(NCCL_INIT, "Parsed SCCL Algorithm %s successfully.", path);
+
+        int regIndex = comm->nScclRegistrations++;
+        NCCLCHECK(ncclRealloc(&comm->scclRegistrations, comm->nScclRegistrations));
```

Comment:
why not use the same structure as scclAlgorithms in com? the array is allocated statically.

Reply:
I wanted to avoid inflating SCCL_MAX_NUM_ALGOS, since that's defined to a fairly low number and gets passed to the device too. The idea is that later on you could have multiple registrations for the same algorithm with different size ranges and different protocol for each range.